File: src/main/java/dev/emortal/tnt/TNT.java
package dev.emortal.tnt;

import com.github.luben.zstd.Zstd;
import dev.emortal.tnt.source.TNTSource;
import net.minestom.server.MinecraftServer;
import net.minestom.server.instance.*;
import net.minestom.server.instance.block.Block;
import net.minestom.server.utils.binary.BinaryWriter;
import org.jglrxavpok.hephaistos.nbt.NBTCompound;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.concurrent.CountDownLatch;
import java.util.stream.Collectors;

public class TNT {

    private static Logger LOGGER = LoggerFactory.getLogger("TNT");


    private static byte[] convertChunk(Chunk chunk) throws IOException {
        BinaryWriter writer = new BinaryWriter();

        writer.writeInt(chunk.getChunkX());
        writer.writeInt(chunk.getChunkZ());
        writer.writeByte((byte) chunk.getMinSection());
        writer.writeByte((byte) chunk.getMaxSection());

        int airSkip = 0;
        boolean needsEnding = false;

        for (int y = chunk.getMinSection() * Chunk.CHUNK_SECTION_SIZE; y < chunk.getMaxSection() * Chunk.CHUNK_SECTION_SIZE; y++) {
            for (int x = 0; x < Chunk.CHUNK_SIZE_X; x++) {
                for (int z = 0; z < Chunk.CHUNK_SIZE_Z; z++) {
                    Block block = chunk.getBlock(x, y, z);

                    // check for the air block state
                    if (block.stateId() == 0) {
                        airSkip++;
                        if (airSkip == 1) {
                            writer.writeShort((short) 0);
                            needsEnding = true;
                        }

                        continue;
                    }
                    if (airSkip > 0) {
                        writer.writeInt(airSkip);
                        needsEnding = false;
                    }

                    airSkip = 0;

                    writer.writeShort(block.stateId());

                    NBTCompound nbt = block.nbt();
                    writer.writeBoolean(block.hasNbt());
                    if (nbt != null) {
                        writer.writeNBT("blockNBT", nbt);
                    }
                }
            }
        }

        // Air skip sometimes isn't written, maybe there is a cleaner way?
        if (needsEnding) {
            writer.writeInt(airSkip);
        }

        for (Section section : chunk.getSections()) {
            writer.writeByteArray(section.getBlockLight());
            writer.writeByteArray(section.getSkyLight());
        }

        byte[] bytes = writer.toByteArray();

        writer.close();
        writer.flush();

        return bytes;
    }

    public static void convertAnvilToTNT(Path anvilPath, TNTSource source) throws IOException, InterruptedException {
        InstanceManager im = MinecraftServer.getInstanceManager();

        Set<Path> mcas = Files.list(anvilPath.resolve("region")).collect(Collectors.toSet());

        InstanceContainer convertInstance = im.createInstanceContainer();
        AnvilLoader loader = new AnvilLoader(anvilPath);
        convertInstance.setChunkLoader(loader);

        CountDownLatch cdl = new CountDownLatch(mcas.size() * 32 * 32);

        ArrayList<byte[]> convertedChunks = new ArrayList<>();

        for (Path mca : mcas) {
            String[] args = mca.getFileName().toString().split("\\.");
            int rX = Integer.parseInt(args[1]);
            int rZ = Integer.parseInt(args[2]);

            for (int x = rX * 32; x < rX * 32 + 32; x++) {
                for (int z = rZ * 32; z < rZ * 32 + 32; z++) {
                    convertInstance.loadChunk(x, z).thenAccept(chunk -> {
                        if (!chunkContainsBlocks(chunk)) {
                            cdl.countDown();
                            return;
                        }

                        byte[] converted = new byte[0];
                        try {
                            converted = convertChunk(chunk);
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                        convertedChunks.add(converted);
                        converted = null;

                        // We're now done with this chunk
                        convertInstance.unloadChunk(chunk);

                        cdl.countDown();
                    });
                }
            }
        }

        cdl.await();

        MinecraftServer.getInstanceManager().unregisterInstance(convertInstance);

        BinaryWriter writer = new BinaryWriter();

        writer.writeInt(convertedChunks.size());
        for (byte[] chunk : convertedChunks) {
            writer.writeBytes(chunk);
        }

        byte[] bytes = writer.toByteArray();
        source.save(bytes);

        writer.close();
        writer.flush();
    }


    public static Collection<Chunk> filterEmptyChunks(Collection<Chunk> chunks) {
        Set<Chunk> newChunks = new HashSet<>();

        for (Chunk chunk : chunks) {
            if (!chunkContainsBlocks(chunk)) continue;

            newChunks.add(chunk);
        }

        return newChunks;
    }

    public static boolean chunkContainsBlocks(Chunk chunk) {
        boolean containsBlocks = false;
        for (Section section : chunk.getSections()) {
            if (section.blockPalette().count() > 0) {
                containsBlocks = true;
                break;
            }
        }
        return containsBlocks;
    }

    public static void convertChunksToTNT(Collection<Chunk> chunks, TNTSource source) throws IOException {
        BinaryWriter writer = new BinaryWriter();

        writer.writeInt(chunks.size());
        for (Chunk chunk : chunks) {
            byte[] converted = new byte[0];
            try {
                converted = convertChunk(chunk);
            } catch (IOException e) {
                e.printStackTrace();
            }

            writer.writeBytes(converted);
        }

        byte[] bytes = writer.toByteArray();
        source.save(bytes);

        writer.close();
        writer.flush();
    }

}

File: src/main/kotlin/dev/emortal/tnt/TNT.kt
package dev.emortal.tnt

import com.github.jinahya.bit.io.BitOutputAdapter
import com.github.jinahya.bit.io.StreamByteOutput
import com.github.luben.zstd.Zstd
import dev.emortal.tnt.source.TNTSource
import net.minestom.server.MinecraftServer
import net.minestom.server.coordinate.Vec
import net.minestom.server.instance.Chunk
import net.minestom.server.instance.block.Block
import net.minestom.server.utils.MathUtils
import net.minestom.server.utils.binary.BinaryWriter
import net.minestom.server.utils.chunk.ChunkUtils
import org.jglrxavpok.hephaistos.nbt.NBTCompound
import org.slf4j.LoggerFactory
import java.nio.file.Files
import java.nio.file.Path
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CountDownLatch
import java.util.stream.Collectors
import kotlin.io.path.nameWithoutExtension

object TNT {

    private val LOGGER = LoggerFactory.getLogger(TNT::class.java)

    /**
     * Saves a .TNT world from chunks
     */
    fun createTNTFile(chunks: Collection<Chunk>, tntSource: TNTSource, saveLights: Boolean = true) {
        val binaryWriter = BinaryWriter()

        val writer = BitOutputAdapter.from(StreamByteOutput.from(binaryWriter))

        // Write settings
        writer.writeBoolean(saveLights)

        writer.writeInt32(chunks.size)

        val paletteBlocks = ArrayList<String>()
        val paletteIndexes = ArrayList<Int>(chunks.size * 16 * 384 * 16)

        chunks.forEach {

            val indexToNBT = ConcurrentHashMap<Int, NBTCompound>()

            writer.writeInt32(it.chunkX)
            writer.writeInt32(it.chunkZ)

            writer.writeByte8(it.minSection.toByte())
            writer.writeByte8(it.maxSection.toByte())

            if (saveLights) {
                // Write lighting
                it.sections.forEach { section ->
                    val blockLight = section.blockLight
                    writer.writeInt32(blockLight.size)
                    blockLight.forEach {
                        writer.writeByte8(it)
                    }

                    val skyLight = section.skyLight
                    writer.writeInt32(skyLight.size)
                    skyLight.forEach {
                        writer.writeByte8(it)
                    }
                }
            }

            for (x in 0 until Chunk.CHUNK_SIZE_X) {
                for (y in -64 until 320) {
                    for (z in 0 until Chunk.CHUNK_SIZE_X) {
                        val pos = Vec(x.toDouble(), y.toDouble(), z.toDouble())
                        val block = it.getBlock(pos)
                        val blockString = stringFromBlock(block)

                        val index = paletteBlocks.indexOf(blockString)
                        if (index == -1) {
                            paletteBlocks.add(blockString)
                            paletteIndexes.add(paletteBlocks.size - 1)
                        } else {
                            paletteIndexes.add(index)
                        }

                        val nbt = block.nbt()
                        if (nbt != null) indexToNBT[ChunkUtils.getBlockIndex(x, y, z)] = nbt
                    }
                }
            }

            // Write NBT
            writer.writeInt32(indexToNBT.size)
            indexToNBT.forEach {
                binaryWriter.writeInt(it.key)
                binaryWriter.writeNBT("blockNBT", it.value)
            }

            // TODO: Global palette option
            // Write blocks then clear for next chunk
            val maxBitsForState = MathUtils.bitsToRepresent(paletteBlocks.size) + 1
            writer.writeInt32(maxBitsForState)

            writer.writeInt32(paletteBlocks.size)
            LOGGER.info("palette size: ${paletteBlocks.size}")
            paletteBlocks.forEach {
                //binaryWriter.writeShort(it)
                binaryWriter.writeSizedString(it)
            }

            writer.writeInt32(paletteIndexes.size)
            //LOGGER.info("Palette blocks indexes ${paletteIndexes.size}")
            paletteIndexes.forEach {
                writer.writeInt(maxBitsForState, it)
                //binaryWriter.writeInt(it)
            }
            paletteBlocks.clear()
            paletteIndexes.clear()
        }

        val bytes: ByteArray = binaryWriter.toByteArray()
        val compressed = Zstd.compress(bytes)

        tntSource.save(compressed)

        writer.close()
        writer.flush()
        binaryWriter.close()
        binaryWriter.flush()
    }

    /**
     * Converts an anvil folder to a TNT file
     *
     * Source support should be added
     */
    fun convertAnvilToTNT(pathToAnvil: Path, tntSaveSource: TNTSource) {
        val instanceManager = MinecraftServer.getInstanceManager()

        val mcaFiles = Files.list(pathToAnvil.resolve("region")).collect(
            Collectors.toSet())

        val convertInstance = instanceManager.createInstanceContainer()
        val loader = ConversionAnvilLoader(pathToAnvil)
        convertInstance.chunkLoader = loader

        val countDownLatch = CountDownLatch((mcaFiles.size) * 32 * 32) // each MCA file contains 32 chunks on each axis
        val chunks: MutableSet<Chunk> = ConcurrentHashMap.newKeySet()

        mcaFiles.forEach {
            val args = it.nameWithoutExtension.split(".").takeLast(2)
            val rX = args[0].toInt()
            val rZ = args[1].toInt()

            for (x in rX * 32 until rX * 32 + 32) {
                for (z in rZ * 32 until rZ * 32 + 32) {
                    convertInstance.loadChunk(x, z).thenAcceptAsync {
                        // Ignore chunks that contain no blocks
                        if (it.sections.any { it.blockPalette().count() > 0 }) chunks.add(it)

                        countDownLatch.countDown()
                    }
                }
            }
        }

        val before = System.nanoTime()
        countDownLatch.await()
        println("Took ${(System.nanoTime() - before) / 1_000_000}ms to convert")

        // TODO: make source independant
        createTNTFile(chunks, tntSaveSource)

        instanceManager.unregisterInstance(convertInstance)
    }

    fun stringFromBlock(block: Block): String {
        val stringBuilder = StringBuilder()
        stringBuilder.append(block.name())

        if (block.properties().isNotEmpty()) {
            stringBuilder.append("[")
            var i = 0
            block.properties().forEach {
                if (i != 0) stringBuilder.append(",")
                stringBuilder.append(it.key)
                stringBuilder.append("=")
                stringBuilder.append(it.value)
                i++
            }
            stringBuilder.append("]")
        }

        return stringBuilder.toString()
    }

    fun blockFromString(string: String): Block {
        if (string.contains("[")) {
            val args = string.split("[")
            val namespace = args[0]
            val properties = args[1]
                .replace("]", "")
                .split(",")
                .associate {
                    val args = it.split("=")
                    args[0] to args[1]
                }

            return Block.fromNamespaceId(namespace)!!.withProperties(properties)
        } else {
            return Block.fromNamespaceId(string)!!
        }
    }

}